src/make_watertight/Arc.cpp

```diff
@@ -0,0 +1,713 @@
+    if (moab::MB_SUCCESS != result) {
+      std::cout << "result=" << result << " could not get adjacencies of edge" << std::endl;
+      return result;
+    }
+
+    // get adjacent edges, but only keep the edges adjacent to both verts
+    moab::Range adj_edges;
+    result = MBI()->get_adjacencies(two_verts, 1, false, adj_edges, moab::Interface::INTERSECT);
+    assert(moab::MB_SUCCESS == result);
+
+    // if any other edges exist, they are opposite the original edge and should be
+    // removed from the skin
+    if (1 < adj_edges.size()) {
+      if (debug) {
+        std::cout << adj_edges.size()
+                  << " opposite edges will be removed from the surface skin "
+                  << adj_edges[0] << " " << adj_edges[1] << std::endl;
+      }
+      edges = subtract(edges, adj_edges);
+      result = MBI()->delete_entities(adj_edges);
+      assert(moab::MB_SUCCESS == result);
+      i--;
+    }
+  }
+  return moab::MB_SUCCESS;
+}
+
+moab::ErrorCode Arc::remove_opposite_pairs_of_edges_fast(moab::Range& edges, const bool debug) {
+  // special case
+  moab::ErrorCode rval;
+  if (1 == edges.size()) {
+    std::cout << "cannot remove pairs: only one input edge" << std::endl;
+    return moab::MB_FAILURE;
+  }
+
+  // populate edge array, used only for searching
+  unsigned n_orig_edges = edges.size();
+  edge* my_edges = new edge[n_orig_edges];
+  unsigned j = 0;
+  for (moab::Range::const_iterator i = edges.begin(); i != edges.end(); ++i) {
+    // get the endpoints of the edge
+    const moab::EntityHandle* endpts;
+    int n_verts;
+    rval = MBI()->get_connectivity(*i, endpts, n_verts);
+    if (moab::MB_SUCCESS != rval || 2 != n_verts) {
+      MB_CHK_SET_ERR(moab::MB_FAILURE, "could not get connectivity");
+    }
+
+    // store the edges
+    my_edges[j].edge = *i;
+    my_edges[j].v0   = endpts[0];
+    my_edges[j].v1   = endpts[1];
+
+    // sort edge by handle
+    if (my_edges[j].v1 < my_edges[j].v0) {
+      moab::EntityHandle temp = my_edges[j].v0;
+      my_edges[j].v0 = my_edges[j].v1;
+      my_edges[j].v1 = temp;
+    }
+    ++j;
+  }
+
+  // sort edge array
+  qsort(my_edges, n_orig_edges, sizeof(struct edge), compare_edge);
+
+  // find duplicate edges
+  j = 0;
+  moab::Range duplicate_edges;
+  for (unsigned i = 1; i < n_orig_edges; ++i) {
+    // delete edge if a match exists
+    if (my_edges[j].v0 == my_edges[i].v0 && my_edges[j].v1 == my_edges[i].v1) {
+      duplicate_edges.insert(my_edges[j].edge);
+      // find any remaining matches
+      while (my_edges[j].v0 == my_edges[i].v0 &&
+             my_edges[j].v1 == my_edges[i].v1 &&
+             i < n_orig_edges) {
+        duplicate_edges.insert(my_edges[i].edge);
+        ++i;
+      }
+      // delete the matches
+      edges = subtract(edges, duplicate_edges);
+      rval = MBI()->delete_entities(duplicate_edges);
+      if (moab::MB_SUCCESS != rval) {
+        delete[] my_edges;
```

Comment:
Rather than clearing deleting this value before every return point, I recommend using a RAII approach (smart pointer) to manage `my_edges` [1].

[1] https://en.cppreference.com/w/cpp/language/raii

Reply:
Definitely a good call. Will update to use a vector of structs instead.